compiler/rustc_codegen_ssa/src/back/link.rs

```diff
@@ -2405,6 +2438,34 @@ fn add_upstream_rust_crates<'a>(
                 // external build system already has the native dependencies defined, and it
                 // will provide them to the linker itself.
                 if sess.opts.unstable_opts.link_native_libraries {
+                    if sess.opts.unstable_opts.separate_native_rlib_dependencies {
+                        // if rlib contains native libs as archieves, unpack them to tmpdir
+                        let archive_map = unsafe {
+                            Mmap::map(File::open(&src.rlib.as_ref().unwrap().0).unwrap()).unwrap()
+                        };
+                        let archive =
+                            object::read::archive::ArchiveFile::parse(&*archive_map).unwrap();
```

Comment:
I think this will need to use an archive reader provided by the codegen backend. Object may not be able to parse the archive format on all targets. For example rust-gpu uses tar instead of an ar variant as archive format.

Reply:
Isn't it the same case as in https://github.com/rust-lang/rust/pull/83640?
`rustc_codegen_ssa` already uses `ArchiveFile::parse` to read stuff from rlibs.

Reply:
The default MetadataLoader uses object for loading crate metadata from rlibs and dylibs since #83640. It doesn't support reading arbitrary archive members however. It can be replaced by the codegen backend.

Reply:
Ah, I didn't notice, `fn get_rlib_metadata` in `rustc_codegen_ssa` is just a default implementation.
Then this archive extraction can also be an overridable default, but in-tree backends `rustc_codegen_(llvm,gcc,cranelift)` don't need to override it.

Reply:
Exactly!